feat(curriculum): add Projectile Trajectory project to SciComPy (#54839)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/07-scientific-computing-with-python/learn-encapsulation-by-building-a-projectile-trajectory-calculator/6633c06601c081735063b528.md

    @speed.setter
    def speed(self, s):
       self.__speed = s
    
    def __repr__(self):
        return f'{self.__class__}({self.speed}, {self.height}, {self.angle})'

class Graph:
    __slots__ = ('__coordinates')

    def __init__(self, coord):
        self.__coordinates = coord

    def __repr__(self):
        return f"Graph({self.__coordinates})"

    def create_coordinates_table(self):
        table = '\n  x      y\n'
        for x, y in self.__coordinates:
            table += f'{x:>3}{round(y, 2):>7.2f}\n'

        return table

    def create_trajectory(self):
        rounded_coord = [(x, round(y)) for x, y in self.__coordinates]
        rows = max([y for x, y in rounded_coord]) + 1
        columns = max([x for x, y in rounded_coord]) + 1
        graph_matrix = [
            [PROJECTILE if (x, y) in rounded_coord else " " for x in range(columns)]
            for y in range(rows, -1, -1)
        ]
        graph_rows = ["⊣" + "".join(row) for row in graph_matrix] + [
            " " + "⊤" * columns
        ]
        graph = "\n" + "\n".join(graph_rows) + "\n"
        return graph

--fcc-editable-region--

--fcc-editable-region--

ball = Projectile(45, 45, 45)

> print(ball)

Reviewer: Probably we want to remove those lines some point before this step.

---
id: 6633c06601c081735063b528
title: Step 23
challengeType: 20
dashedName: step-23
---

# --description--

You have built a projectile trajectory calculator.

> Now to conclude, create a little utility function called `projectile_helper` outside the classes that takes in the desired values for speed, height and angle and prints it to the terminal in sequence, the details of the projectile, the table of coordinates and the graph of the trajectory.

Reviewer: ```suggestion
Now to conclude, outside the classes, create a little utility function called `projectile_helper` that takes in the desired values for speed, height and angle and prints it to the terminal in sequence, the details of the projectile, the table of coordinates and the graph of the trajectory.
```

> Now to conclude, outside the classes, create a little utility function called `projectile_helper` that takes in the desired values for speed, height and angle and prints it to the terminal in sequence, the details of the projectile, the table of coordinates and the graph of the trajectory.

Reviewer: ```suggestion
Now to conclude, modify the code you wrote outside the classes, and incorporate it into a little utility function called `projectile_helper` that takes in the desired values for speed, height and angle and prints to the terminal in sequence, the details of the projectile, the table of coordinates and the graph of the trajectory.
```

    def speed(self, s):
       self.__speed = s
    
    def __repr__(self):
        return f'{self.__class__}({self.speed}, {self.height}, {self.angle})'

class Graph:
    __slots__ = ('__coordinates')

    def __init__(self, coord):
        self.__coordinates = coord

    def __repr__(self):
        return f"Graph({self.__coordinates})"

    def create_coordinates_table(self):
        table = '\n  x      y\n'
        for x, y in self.__coordinates:
            table += f'{x:>3}{round(y, 2):>7.2f}\n'

        return table

    def create_trajectory(self):

        rounded_coords = [(round(x), round(y)) for x, y in self.__coordinates]

        x_max = max(rounded_coords, key=lambda i: i[0])[0]
        y_max = max(rounded_coords, key=lambda j: j[1])[1]

        matrix_list = [[" " for _ in range(x_max + 1)] for _ in range(y_max + 1)]

        for x, y in rounded_coords:
            matrix_list[-1 - y][x] = PROJECTILE

        matrix = ["".join(line) for line in matrix_list]

        matrix_axes = [y_axis_tick + row for row in matrix]
        matrix_axes.append(" " + x_axis_tick * (len(matrix[0])))

        graph = "\n" + "\n".join(matrix_axes) + "\n"

        return graph

> ball = Projectile(10, 3, 45)
> print(ball)
> coordinates = ball.calculate_all_coordinates()
> graph = Graph(coordinates)
> print(graph.create_trajectory())
> 
> --fcc-editable-region--
> 
> --fcc-editable-region--

Reviewer: ```suggestion
--fcc-editable-region--
ball = Projectile(10, 3, 45)
print(ball)
coordinates = ball.calculate_all_coordinates()
graph = Graph(coordinates)
print(graph.create_trajectory())

--fcc-editable-region--
```

Reviewer 2: TODO: change solution

---
id: 6633c06601c081735063b528
title: Step 21
challengeType: 20
dashedName: step-21
---

# --description--

You have built a projectile trajectory calculator.

> Now to conclude, create a little utility function called`projectile_helper` outside the classes that takes in the desired values for speed, height and angle and print to the terminal in sequence, the details of the projectile, the table of coordinates and the graph of the trajectory.

Reviewer 3: ```suggestion
Now to conclude, create a little utility function called `projectile_helper` outside the classes that takes in the desired values for speed, height and angle and prints it to the terminal in sequence, the details of the projectile, the table of coordinates and the graph of the trajectory.
```

c = bullet.calculate_all_coordinates()
g = Graph(c)
assert __result[1] == g.create_coordinates_table()
assert __result[2] == g.create_trajectory()
        `)
    }
})
```

You should call the `projectile_helper` function at least once.

```js
({
    test: () => runPython(`
    assert len(_Node(_code).find_calls('projectile_helper')) > 0
    `)
})
```

You should not have variables or print calls in the global scope.

```js
({
    test: () => runPython(`
    assert _Node(_code).find_calls('print') == [], "found print"
    assert not _Node(_code).has_variable('graph'), "found graph"
    assert not _Node(_code).has_variable('ball'), "found ball"
    assert not _Node(_code).has_variable('coordinates'), "found coordinates"
    `)
})
```

# --seed--

## --seed-contents--

```py
import math

GRAVITATIONAL_ACCELERATION = 9.81
PROJECTILE = "∙"
x_axis_tick = "T"
y_axis_tick = "⊣"

class Projectile:
    __slots__ = ('__speed', '__height', '__angle')

    def __init__(self, speed, height, angle):
        self.__speed = speed
        self.__height = height
        self.__angle = math.radians(angle)
        
    def __str__(self):
        return f'''
Projectile details:

> speed: {self.__speed} m/s
> height: {self.__height} m
> angle: {round(math.degrees(self.__angle))}°

Reviewer: ```suggestion
speed: {self.speed} m/s
height: {self.height} m
angle: {self.angle}°
```

        x_max = max(rounded_coords, key=lambda i: i[0])[0]
        y_max = max(rounded_coords, key=lambda j: j[1])[1]

        matrix_list = [[" " for _ in range(x_max + 1)] for _ in range(y_max + 1)]

        for x, y in rounded_coords:
            matrix_list[-1 - y][x] = PROJECTILE

        matrix = ["".join(line) for line in matrix_list]

        matrix_axes = [y_axis_tick + row for row in matrix]
        matrix_axes.append(" " + x_axis_tick * (len(matrix[0])))

        graph = "\n" + "\n".join(matrix_axes) + "\n"

        return graph

--fcc-editable-region--
ball = Projectile(10, 3, 45)
print(ball)
coordinates = ball.calculate_all_coordinates()
graph = Graph(coordinates)
print(graph.create_trajectory())

--fcc-editable-region--
```

# --solutions--

```py
import math

GRAVITATIONAL_ACCELERATION = 9.81
PROJECTILE = "∙"
x_axis_tick = "T"
y_axis_tick = "⊣"

class Projectile:
    __slots__ = ('__speed', '__height', '__angle')

    def __init__(self, speed, height, angle):
        self.__speed = speed
        self.__height = height
        self.__angle = math.radians(angle)
        
    def __str__(self):
        return f'''
Projectile details:

> speed: {self.__speed} m/s
> height: {self.__height} m
> angle: {round(math.degrees(self.__angle))}°

Reviewer: ```suggestion
speed: {self.speed} m/s
height: {self.height} m
angle: {self.angle}°
```
